File: Problems/No.101_Symmetric_Tree/symmetric_tree_recursive.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def isSymmetric(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        
        def checkSymmetric(p, q):
            
            if (p!=None and q!=None):
                return checkSymmetric(p.left, q.right) and (p.val==q.val) and checkSymmetric(p.right, q.left)
            elif (p==None and q==None):
                return True
            else:
                return False
        
        return checkSymmetric(root.left, root.right)
        

File: Problems/No.101_Symmetric_Tree/symmetric_tree_recursive.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
    bool checkSymmetric(TreeNode* p, TreeNode* q)  {
        
        if (p!=NULL and q!=NULL) return checkSymmetric(p->left, q->right) and (p->val==q->val) and checkSymmetric(p->right, q->left);
        else if (p==NULL and q==NULL) return true;
        else return false;
        
    }
    
    bool isSymmetric(TreeNode* root) {
        
        return checkSymmetric(root->left, root->right);
        
    }
};
